File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Query.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.firestore.conformance.model;

import androidx.annotation.NonNull;
import com.google.auto.value.AutoValue;
import java.util.Collections;
import java.util.List;

/** A query call stack to Firestore. */
@AutoValue
public abstract class Query {

  /** Name of the collection to query from. */
  public abstract String getCollection();

  /** Filters. */
  public abstract List<QueryFilter> getFilters();

  /** Builder for {@link Query}. */
  @AutoValue.Builder
  public abstract static class Builder {
    @NonNull
    public static Query.Builder builder() {
      return new AutoValue_Query.Builder().setFilters(Collections.emptyList());
    }

    @NonNull
    public abstract Query.Builder setCollection(@NonNull String collection);

    @NonNull
    public abstract Query.Builder setFilters(@NonNull List<QueryFilter> filters);

    @NonNull
    public abstract Query build();
  }
}

File: firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/Query.kt
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.firebase.dataconnect

import java.util.Objects
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.SerializationStrategy

public class Query<Response, Variables>
internal constructor(
  dataConnect: FirebaseDataConnect,
  operationName: String,
  responseDeserializer: DeserializationStrategy<Response>,
  variablesSerializer: SerializationStrategy<Variables>,
) :
  Reference<Response, Variables>(
    dataConnect = dataConnect,
    operationName = operationName,
    responseDeserializer = responseDeserializer,
    variablesSerializer = variablesSerializer,
  ) {
  override suspend fun execute(variables: Variables): DataConnectQueryResult<Response, Variables> =
    dataConnect.lazyQueryManager.get().execute(this, variables)

  public fun subscribe(variables: Variables): QuerySubscription<Response, Variables> =
    QuerySubscription(this, variables)

  override fun hashCode(): Int = Objects.hash("Query", super.hashCode())

  override fun equals(other: Any?): Boolean = (other is Query<*, *>) && super.equals(other)

  override fun toString(): String =
    "Query(" +
      "dataConnect=$dataConnect, " +
      "operationName=$operationName, " +
      "responseDeserializer=$responseDeserializer, " +
      "variablesSerializer=$variablesSerializer" +
      ")"

  public fun <NewResponse> withResponseDeserializer(
    newResponseDeserializer: DeserializationStrategy<NewResponse>
  ): Query<NewResponse, Variables> =
    Query(
      dataConnect = dataConnect,
      operationName = operationName,
      responseDeserializer = newResponseDeserializer,
      variablesSerializer = variablesSerializer,
    )

  public fun <NewVariables> withVariablesSerializer(
    newVariablesSerializer: SerializationStrategy<NewVariables>
  ): Query<Response, NewVariables> =
    Query(
      dataConnect = dataConnect,
      operationName = operationName,
      responseDeserializer = responseDeserializer,
      variablesSerializer = newVariablesSerializer,
    )
}
